src/java/org/apache/cassandra/index/sai/disk/v1/bbtree/BlockBalancedTreeWriter.java

```diff
@@ -0,0 +1,779 @@
+        private final long[] leafRowIDs = new long[maxPointsInLeafNode];
+        private final RowIDAndIndex[] rowIDAndIndexes = new RowIDAndIndex[maxPointsInLeafNode];
+        private final int[] orderIndex = new int[maxPointsInLeafNode];
+        private final Callback callback;
+        private final GrowableByteArrayDataOutput scratchOut = new GrowableByteArrayDataOutput(32 * 1024);
+        private final GrowableByteArrayDataOutput scratchOut2 = new GrowableByteArrayDataOutput(2 * 1024);
+        private final byte[] lastPackedValue = new byte[bytesPerValue];
+
+        private long valueCount;
+        private int leafValueCount;
+        private long lastRowID;
+
+        TreeWriter(IndexOutput treeOutput, Callback callback)
+        {
+            assert callback != null : "Callback cannot be null in TreeWriter";
+
+            this.treeOutput = treeOutput;
+            this.callback = callback;
+
+            for (int x = 0; x < rowIDAndIndexes.length; x++)
+            {
+                rowIDAndIndexes[x] = new RowIDAndIndex();
+            }
+        }
+
+        void add(byte[] packedValue, long docID) throws IOException
+        {
+            if (DEBUG)
+                valueInOrder(valueCount + leafValueCount, lastPackedValue, packedValue, 0, docID, lastRowID);
+
+            System.arraycopy(packedValue, 0, leafValues, leafValueCount * bytesPerValue, bytesPerValue);
+            leafRowIDs[leafValueCount] = docID;
+            leafValueCount++;
+
+            if (leafValueCount == maxPointsInLeafNode)
+            {
+                // We write a block once we hit exactly the max count
+                writeLeafBlock();
+                leafValueCount = 0;
+            }
+
+            if (DEBUG)
+                if ((lastRowID = docID) < 0)
+                    throw new AssertionError("document id must be >= 0; got " + docID);
+        }
+
+        /**
+         * Write a leaf block if we have unwritten values and return the total number of values added
+         */
+        public long finish() throws IOException
+        {
+            if (leafValueCount > 0)
+                writeLeafBlock();
+
+            return valueCount;
+        }
+
+        private void writeLeafBlock() throws IOException
+        {
+            assert leafValueCount != 0;
+            if (valueCount == 0)
+            {
+                System.arraycopy(leafValues, 0, minPackedValue, 0, bytesPerValue);
+            }
+            System.arraycopy(leafValues, (leafValueCount - 1) * bytesPerValue, maxPackedValue, 0, bytesPerValue);
+
+            valueCount += leafValueCount;
+
+            if (leafBlockFilePointer.size() > 0)
+            {
+                // Save the first (minimum) value in each leaf block except the first, to build the split value index in the end:
+                leafBlockStartValues.add(ArrayUtil.copyOfSubArray(leafValues, 0, bytesPerValue));
```

Comment:
I was having trouble hitting this w/ the tests, and then I realized that most of them don't actually full up a leaf. For instance, I took `NumericIndexWriterTest#shouldFlushFromMemtable()` and boosted to `final int maxSegmentRowId = NumericIndexWriter.MAX_POINTS_IN_LEAF_NODE * 3;`. This and all the other code for rotating the tree, etc. is then hit.

Are there actually other tests that write more than a single leaf, or do we need to pick some places like this ^ to extend coverage?

Reply:
I think we need to boost the testing on this. I found the same thing when I was doing the initial analysis of the internals and had to create an extended test to hit some of these code paths. Now I have a better understanding of these internals I'll try and put something together that correctly hits and tests these paths.

Reply:
`BlockBalancedTreeReaderTest` looks to be the test that hits most of the conditions. It definitely hits this one and does create multi-leaf trees.